OpenRA.Mods.RA2/Orders/EnterGarrisonTargeter.cs

```diff
@@ -0,0 +1,77 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Mods.RA2.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.RA2.Orders
+{
+    public class EnterGarrisonTargeter<T> : UnitOrderTargeter where T : ITraitInfo
+    {
+        readonly OpenRA.Mods.RA2.Traits.AlternateTransportsMode mode;
+        readonly Func<Actor, Actor, bool> canTarget;
+        readonly Func<Actor, Actor, bool> useEnterCursor;
+
+        public EnterGarrisonTargeter(string order, int priority,
+            Func<Actor, Actor, bool> canTarget, Func<Actor, Actor, bool> useEnterCursor, OpenRA.Mods.RA2.Traits.AlternateTransportsMode mode)
+            : base(order, priority, "enter", true, true)
+        {
+            this.canTarget = canTarget;
+            this.useEnterCursor = useEnterCursor;
+            this.mode = mode;
+        }
+
+        public override bool CanTargetActor(Actor self, Actor target, TargetModifiers modifiers, ref string cursor)
+        {
+            switch (mode)
+            {
+                case OpenRA.Mods.RA2.Traits.AlternateTransportsMode.None:
+                    return false;
+                case OpenRA.Mods.RA2.Traits.AlternateTransportsMode.Force:
+                    if (!modifiers.HasModifier(TargetModifiers.ForceMove))
+                        return false;
+                    break;
+                case OpenRA.Mods.RA2.Traits.AlternateTransportsMode.Default:
+                    if (modifiers.HasModifier(TargetModifiers.ForceMove))
+                        return false;
+                    break;
+                case OpenRA.Mods.RA2.Traits.AlternateTransportsMode.Always:
+                    break;
+            }
+
+            if ((target.Owner.PlayerName == "Creeps" || target.Owner.PlayerName == "Neutral" || self.Owner.IsAlliedWith(target.Owner)) && target.Info.HasTraitInfo<T>())
+            {
+                cursor = useEnterCursor(self, target) ? "enter" : "enter-blocked";
+                return true;
+            }
+
+            if (!self.Owner.IsAlliedWith(target.Owner) || !target.Info.HasTraitInfo<T>() || !canTarget(self, target) || target.Owner.PlayerName == "Creeps")
+                return false;
+
+            cursor = useEnterCursor(self, target) ? "enter" : "enter-blocked";
+            return true;
+        }
+
+        public override bool CanTargetFrozenActor(Actor self, FrozenActor target, TargetModifiers modifiers, ref string cursor)
```

Comment:
Wasn't CanTargetFrozenActor dropped completely lately? I'm not that familiar with the frozen-actor rework of the past 2-3 weeks.

Reply:
OpenRA.Mods.Common.Orders -> UnitOrderTargeter.cs still requires CanTargetFrozenActor to be implemented. 
But I did cleanup the code here, and added the AlternateGarrisoningMode logic to the function. Now you can hold alt over a frozen actor to get them to move there without entering the structure.